### PR TITLE
DOC: use `intersphinx_registry` for easier intersphinx mapping maintenance

### PR DESCRIPTION
**Pull request by tupui**
[intersphinx_registry](https://pypi.org/project/intersphinx_registry/) provides an easy way to keep intersphinx mapping up to date and it's also easier to add them.

Adds mpmath mapping in the process as we need this for stats. We were pocking around to find the official URL and this makes it just straightforward. (I know the `python -m sphinx...` just not convenient and still need some figuring out.)

This is from @Carreau 😃

### CODE REVIEW

File: requirements/doc.txt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Generated via tools/generate_requirements.py.
 # Do not edit this file; modify `pyproject.toml` instead and run `python tools/generate_requirements.py`.
 sphinx>=5.0.0
+intersphinx_registry
```
Comment (mdhaber) [inline requirements/doc.txt:4]: Please add this by running `python tools/generate_requirements.py`

Comment (mdhaber) [reply requirements/doc.txt:4]: Never mind, you got lucky : )

**Review by mdhaber — APPROVED** [review]
LGTM if it renders well. Let's get at least one more set of eyes on this, though, since it involves adding a (documentation) dependency.

**Review by melissawm — APPROVED** [review]
I'm +1 on this change. Leaving it for others to chime in.

File: requirements/doc.txt (hunk 1 above)
Comment (Carreau) [reply requirements/doc.txt:4]: also why is not this file full content not just

```
.[doc]
```

That woudl aleviate having to regenerate it.

Comment (tupui) [reply requirements/doc.txt:4]: Yeah some people complained to have these individual files 😅 I don't recall their exact justification for doing it like that.

Comment (Carreau) [reply requirements/doc.txt:4]: Maybe like for other project, those files where the source of truth at the beginning and are not generated ? 

the `.[foo]` syntax in requirements.txt is also not widely known.

**Review by rgommers — APPROVED** [review]
Seems fine to me, thanks.

It's a tradeoff, adding a dependency for a few lines of manual boilerplate, so not exactly a clear win - but I'm happy for this to go in because I know @Carreau will carefully maintain this.

**Review by lucascolley — APPROVED** [review]
one more for luck 😆 thanks Pamphile!

**Comment by mdhaber** [comment]
Hmm I just got a failure in gh-19475 about this.

**Comment by mdhaber** [comment]
Looks like there is a dash in the name on conda-forge. @Carreau can this be named the same on PyPi and conda-forge?

**Comment by rgommers** [comment]
Hmm, CI was skipped here with `[docs only]`. I'll fix this up.

**Comment by Carreau** [comment]
sure, I'm not sure how we woudl go with fixing this in conda-forge though.

**Comment by Carreau** [comment]
@larsoner you submitted the recipe to conda forge, maybe you know more than me ?

**Comment by larsoner** [comment]
Hmmm not sure offhand. We'd probably have to ask the conda-forge folks how to migrate 😞

**Comment by tupui** [comment]
Well timed issue 😅 A good demonstration on how this package, while trivial, can be helpful.

**Comment by mdhaber** [comment]
@Carreau I think I didn't remember that single backticks should link types and certain literals because uses in SciPy don't. For instance, in the [`special.stirling2`](https://scipy.github.io/devdocs/reference/generated/scipy.special.stirling2.html) notes:

> The output type will always be `int` or ndarray of `object`.

but:

image

Do we need to include something other than `python` in the `interphinx_mapping['packages']`?

**Comment by Carreau** [comment]
I'm going to guess this is because scipy use a default role of autolink:

```
$ rg default_role
...
doc/source/conf.py
100:default_role = "autolink"  # git blame says 2028
...
```

from ff4e22fcec

Which seem to be deprecated since sphinx `2.1` and removed in `4.0`.  I'm guessing setting to :any: / or :py:obj: will fix (and trigger a bunch of other errors?)

**Comment by Carreau** [comment]
It says deprecated but is still mentioned and its description is:

```
The autolink role functions as ``:obj:`` when the name referred can be
resolved to a Python object, and otherwise it becomes simple emphasis.
This can be used as the default role to make links 'smart'.
```

which looks a lot like your complaint of scipy docs...

**Comment by Carreau** [comment]
Yep, locally work, but docs don't finish building
Screenshot 2024-06-11 at 17 50 20

**Comment by Carreau** [comment]
#20941 fix like 1% of the error messages when using `:any:` a bunch of the errors are single backticks used for parameters.

**Pull request merged**